src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py

```diff
@@ -0,0 +1,196 @@
+        assert temp_decrease <= 0.0
+        return abs(temp_decrease)
+
+    def _calc_temp_increase_K(self, time_slot: DateTime, energy_kWh: float) -> float:
+        storage_temp = self._energy_to_storage_temp(energy_kWh, time_slot)
+        current_storage_temp = self.state.get_storage_temp_C(time_slot)
+        if storage_temp <= current_storage_temp:
+            return 0
+        return storage_temp - current_storage_temp
+
+    def _storage_temp_to_energy(self, storage_temp: float, time_slot: DateTime) -> float:
+        if not self._min_temp_C < storage_temp < self._max_temp_C:
+            logger.info(
+                "Storage temp %s cannot exceed min (%s) / max (%s) tank temperatures.",
+                storage_temp, self._min_temp_C, self._max_temp_C)
+            storage_temp = max(min(storage_temp, self._max_temp_C), self._min_temp_C)
+
+        dh_supply_temp = self._water_supply_temp_C.profile[time_slot]
+        dh_return_temp = self._water_return_temp_C.profile[time_slot]
+        current_storage_temp = self.state.get_storage_temp_C(time_slot)
+        m_m3 = self._dh_water_flow_m3.profile[time_slot]
+        m_kg_per_sec = m_m3 * 1000 / 3600
+        q_out = m_kg_per_sec * WATER_SPECIFIC_TEMPERATURE * (dh_supply_temp - dh_return_temp)
+
+        if m_kg_per_sec <= FLOATING_POINT_TOLERANCE:
+            return 0.
+
+        temp_differential_per_sec = (
+                (storage_temp - current_storage_temp) /
+                GlobalConfig.slot_length.total_seconds())
+        q_in = (WATER_DENSITY * WATER_SPECIFIC_TEMPERATURE *
+                self._tank_volume_l * temp_differential_per_sec + q_out)
+        condenser_temp = (q_in / (m_kg_per_sec * WATER_SPECIFIC_TEMPERATURE)) + storage_temp
+        cop = CALIBRATION_COEFFICIENT * (condenser_temp /
+                                         (condenser_temp - GROUND_WATER_TEMPERATURE_C))
+        p_el = q_in / cop
+        energy_kWh = convert_W_to_kWh(p_el, GlobalConfig.slot_length)
+        logger.debug("Calculated maximum electricity demand for heatpump. \n"
+                     "Target Storage Temperature: %s C. \n"
+                     "Current Storage Temperature: %s C. \n"
+                     "District Heating Supply Temperature: %s C. \n"
+                     "District Heating Return Temperature: %s C. \n"
+                     "District Heating Water Flow: %s m3/hour. \n"
+                     "Temperature Differential: %s C/sec. \n"
+                     "Q Out: %s W. Q In: %s W. \n"
+                     "Condenser Temperature: %s C. COP: %s.\n"
+                     "Heatpump Power: %s W.\n"
+                     "Heatpump Energy Consumption %s kWh.",
+                     storage_temp, current_storage_temp, dh_supply_temp, dh_return_temp, m_m3,
+                     temp_differential_per_sec, q_out, q_in, condenser_temp, cop, p_el, energy_kWh)
+        return energy_kWh
+
+    def _energy_to_storage_temp(
```

Comment:
Also here I would like a docstring for later. For example:
```python
"""
Return the temperature of the storage after the provided energy amount is converted into heat.
"""
```

Reply:
Done